File: #1047.py
MOD = 1_000_000_007
def matrix_multiply(A, B):
    n = len(A)
    m = len(B[0])
    result = [[0] * m for _ in range(n)]
    for i in range(n):
        for j in range(m):
            for k in range(len(B)):
                result[i][j] += A[i][k] * B[k][j] % MOD
                result[i][j] %= MOD
    return result

def matrix_power(X, N):
    if N == 1:
        return X
    if N % 2 == 0:
        B = matrix_power(X, N // 2)
        return matrix_multiply(B, B)
    else:
        B = matrix_power(X, N // 2)
        return matrix_multiply(matrix_multiply(B, B), X)

X = [[0, 1] ,[1, 1]]

n = int(input())
for i in range(n):
    x, y, a = map(int, input().split())
    V = [[x],[y]]
    if a < 2:
        print(V[a][0])
        continue
    result = matrix_power(X, a)
    final_result = matrix_multiply(result, V)
    print(final_result[0][0])

File: #1047.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 1'000'000'007;

vector<vector<int>> matrix_multiply(vector<vector<int>> A, vector<vector<int>> B) {
    int n = A.size();
    int m = B[0].size();
    vector<vector<int>> result(n, vector<int>(m, 0));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            for (int k = 0; k < B.size(); ++k) {
                result[i][j] += (1LL*A[i][k] * B[k][j]) % MOD;
                result[i][j] %= MOD;
            }
        }
    }
    return result;
}

vector<vector<int>> matrix_power(vector<vector<int>> X, int N) {
    if (N == 1) {
        return X;
    }
    if (N % 2 == 0) {
        auto B = matrix_power(X, N / 2);
        return matrix_multiply(B, B);
    } else {
        auto B = matrix_power(X, N / 2);
        return matrix_multiply(matrix_multiply(B, B), X);
    }
}

int main() {
    vector<vector<int>> X = {{0, 1}, {1, 1}};
    int n;
    cin >> n;
    for (int i = 0; i < n; ++i) {
        int x, y, a;
        cin >> x >> y >> a;
        vector<vector<int>> V = {{x}, {y}};
        if (a < 2) {
            cout << V[a][0] << endl;
            continue;
        }
        auto result = matrix_power(X, a);
        auto final_result = matrix_multiply(result, V);
        cout << final_result[0][0] << endl;
    }
    return 0;
}
